src/js/video.js

```diff
@@ -96,10 +96,70 @@ function videojs(id, options, ready) {
   }
 
   // Element may have a player attr referring to an already created player instance.
-  // If not, set up a new player and return the instance.
-  return tag.player || Player.players[tag.playerId] || new Player(tag, options, ready);
+  // If so return that otherwise set up a new player below
+  if (tag.player || Player.players[tag.playerId]) {
+    return tag.player || Player.players[tag.playerId];
+  }
+
+  videojs.hooks('presetup').forEach(function(hookFunction) {
+    options = videojs.mergeOptions(options, hookFunction(tag, options, ready));
+  });
+
+  // If not, set up a new player
+  const player = new Player(tag, options, ready);
+
+  videojs.hooks('postsetup').forEach((hookFunction) => hookFunction(player));
+
+  return player;
 }
 
+/**
+ * An Object that contains lifecycle hooks as keys which point to an array
+ * of functions that are run when a lifecycle event is triggered
+ */
+videojs.hooks_ = {};
+
+/**
+ * Get a list of hooks for a specific lifecycle event
+ *
+ * @param {String} type the lifecyle event to get hooks from
+ * @return {Array} an array of hooks, or an empty array if there are none
+ */
+videojs.hooks = function(type) {
```

Comment:
Is hooks a too-generic term? would we want any more hooks on the main videojs object beyond `presetup` and `postsetup`?